### DISCUSSION TITLE
explainer error

### DISCUSSION THREAD

**Discussion by SriATX** [pyg-team/pytorch_geometric#9503] [Q&A]
I am trying to invoke explainer after running anomaly-detection (using PyGOD's DOMINANT). Here is how I am attempting it:

         ` explainer = Explainer(
            model=trained_anomaly_detector.model,
            algorithm=GNNExplainer(epochs=200),
            explanation_type='model',
            node_mask_type='attributes',
            edge_mask_type='object',
            model_config=dict(
                mode='binary_classification',
                task_level='node',
                return_type='probs', 
            ),
        )

        # Generate explanation for the node at index `25`:
        explanation = explainer(specific_splitted_data.x, specific_splitted_data.edge_index, index=25)`

And, the `trained_anomaly_detector` is fully fitted DOMINANT instance: `type of trained anomaly detector:<class 'pygod.detector.dominant.DOMINANT'>`

The error I am seeing is below. I am not sure why the prediction is returning as a **tuple instead of a float**. 

```
Traceback (most recent call last):
  File "/Users/sri/exps/src/main.py", line 325, in <module>
    main(get_args())
  File "/Users/sri/exps/src/main.py", line 317, in main
    explanation = explainer(data.x, data.edge_index, index=25)
  File "/Users/sri/.pyenv/versions/ve_3.10.13/lib/python3.10/site-packages/torch_geometric/explain/explainer.py", line 196, in __call__
    target = self.get_target(prediction)
  File "/Users/sri/.pyenv/versions/ve_3.10.13/lib/python3.10/site-packages/torch_geometric/explain/explainer.py", line 266, in get_target
    return (prediction > 0.5).long().view(-1)
TypeError: '>' not supported between instances of 'tuple' and 'float'
make: *** [run_ad] Error 1

```

For completeness sake, when I run the DOMINANT detector, I run it like this and get the prediction results as expected:
```
        pred, score, prob, conf = self.detector.predict(self.graph_data,
                                                        return_pred=True,
                                                        return_score=True,
                                                        return_prob=True,
                                                        return_conf=True)


        Labels:
        tensor([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
                0, 1, 1, 1, 0, 0, 0, 0])
        
        Raw scores:
        tensor([2.5767, 1.7400, 1.6857, 1.6642, 2.1855, 1.6709, 1.8888, 1.7068, 1.7821,
                1.7520, 1.7857, 2.8096, 1.7170, 2.0270, 1.9223, 1.7018, 1.9239, 1.9717,
                1.6906, 1.7560, 1.7277, 1.7371, 2.3796, 1.7316, 1.9786, 3.2201, 2.2739,
                2.4594, 1.6745, 1.9496, 1.7073, 1.7392])
        
        Probability:
        tensor([0.5766, 0.0000, 0.0000, 0.0000, 0.2718, 0.0000, 0.0405, 0.0000, 0.0000,
                0.0000, 0.0000, 0.7580, 0.0000, 0.1483, 0.0666, 0.0000, 0.0679, 0.1051,
                0.0000, 0.0000, 0.0000, 0.0000, 0.4230, 0.0000, 0.1105, 1.0000, 0.3406,
                0.4852, 0.0000, 0.0879, 0.0000, 0.0000])
        
        Confidence:
        tensor([0.7095, 1.0000, 1.0000, 1.0000, 0.8691, 1.0000, 1.0000, 1.0000, 1.0000,
                1.0000, 1.0000, 0.7095, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
                1.0000, 1.0000, 1.0000, 1.0000, 0.4548, 1.0000, 1.0000, 0.9330, 0.4548,
                0.7095, 1.0000, 1.0000, 1.0000, 1.0000])
```

And, the outlier scores seem to be of the expected datatype (tensor):

```
outlier_scores = self.detector.decision_function(data=self.graph_data)

outlier scores: 
       tensor([2.5767, 1.7400, 1.6857, 1.6642, 2.1855, 1.6709, 1.8888, 1.7068, 1.7821,
               1.7520, 1.7857, 2.8096, 1.7170, 2.0270, 1.9223, 1.7018, 1.9239, 1.9717,
               1.6906, 1.7560, 1.7277, 1.7371, 2.3796, 1.7316, 1.9786, 3.2201, 2.2739,
               2.4594, 1.6745, 1.9496, 1.7073, 1.7392])
```

**Comment by rusty1s** [comment]
It looks like the model returns a `tuple`, which our explainer cannot handle. You may want to wrap the model and only output a single tensor.

**Reply by SriATX** [reply]
@rusty1s Can you please elaborate how to 'wrap' the model to output a single tensor?  Below is my Anomaly Detector class. Can you please point to me where/how this wrapping is done? An eg code will help. Thanks!

`class Anomaly_Detection:
    def __init__(self, graph_dataset=None, detector=DOMINANT, args_dict=None):
        self.graph_data = graph_dataset
        self.args_dict = args_dict
        self.detector = detector

    def train_anomaly_detector(self):
        self.detector = DOMINANT(hid_dim=64, num_layers=4, epoch=20)  # hyperparameters can be set here
        return self.detector.fit(self.graph_data)

    def eval_anomaly_detector(self):
        """
        Ground Truth Y:[False False False False False False False False False False False False
         False False False False False False False False False False False False
         False  True False False False False False False]
        Predicted Scores: tensor([2.5767, 1.7400, 1.6857, 1.6642, 2.1855, 1.6709, 1.8888, 1.7068, 1.7821,
                1.7520, 1.7857, 2.8096, 1.7170, 2.0270, 1.9223, 1.7018, 1.9239, 1.9717,
                1.6906, 1.7560, 1.7277, 1.7371, 2.3796, 1.7316, 1.9786, 3.2201, 2.2739,
                2.4594, 1.6745, 1.9496, 1.7073, 1.7392])
        AUC Score: 1.000
        AP Score: 1.000
        """
        outlier_scores = self.detector.decision_function(data=self.graph_data)
        ground_truth_label = self.graph_data.y.numpy()

        auc = roc_auc_score(ground_truth_label, outlier_scores)
        ap = average_precision_score(ground_truth_label, outlier_scores)
        if self.args_dict['debug']:
            print(f'\n\nGround Truth Y:{ground_truth_label}')
            print(f'Predicted Scores: {outlier_scores}')

            print(f'AUC Score: {auc:.3f}')
            print(f'AP Score: {ap:.3f}')

        # ------------------
        """
        Labels:
        tensor([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
                0, 1, 1, 1, 0, 0, 0, 0])

        Raw scores:
        tensor([2.5767, 1.7400, 1.6857, 1.6642, 2.1855, 1.6709, 1.8888, 1.7068, 1.7821,
                1.7520, 1.7857, 2.8096, 1.7170, 2.0270, 1.9223, 1.7018, 1.9239, 1.9717,
                1.6906, 1.7560, 1.7277, 1.7371, 2.3796, 1.7316, 1.9786, 3.2201, 2.2739,
                2.4594, 1.6745, 1.9496, 1.7073, 1.7392])

        Probability:
        tensor([0.5766, 0.0000, 0.0000, 0.0000, 0.2718, 0.0000, 0.0405, 0.0000, 0.0000,
                0.0000, 0.0000, 0.7580, 0.0000, 0.1483, 0.0666, 0.0000, 0.0679, 0.1051,
                0.0000, 0.0000, 0.0000, 0.0000, 0.4230, 0.0000, 0.1105, 1.0000, 0.3406,
                0.4852, 0.0000, 0.0879, 0.0000, 0.0000])
                
        Confidence:
        tensor([0.7095, 1.0000, 1.0000, 1.0000, 0.8691, 1.0000, 1.0000, 1.0000, 1.0000,
                1.0000, 1.0000, 0.7095, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
                1.0000, 1.0000, 1.0000, 1.0000, 0.4548, 1.0000, 1.0000, 0.9330, 0.4548,
                0.7095, 1.0000, 1.0000, 1.0000, 1.0000])
    
        """
        pred, score, prob, conf = self.detector.predict(self.graph_data,
                                                        return_pred=True,
                                                        return_score=True,
                                                        return_prob=True,
                                                        return_conf=True)
        if self.args_dict['debug']:
            print('\n\nLabels:')
            print(pred)

            print('\nRaw scores:')
            print(score)

            print('\nProbability:')
            print(prob)

            print('\nConfidence:')
            print(conf)
        return ground_truth_label, outlier_scores

`
